text/0000-serialization-format.md

```diff
@@ -0,0 +1,125 @@
+# Serialization Format
+
+## Summary
+
+[summary]: #summary
+
+Definition of a versioned, backwards-compatible and extensible serialization
+format for the state of Cartographer.
+
+This RFC is a small deviation of the previously proposed header format from
+[RFC-0015](https://github.com/googlecartographer/rfcs/blob/master/text/0015-serialization-header.md).
+
+#### Note: This is **NOT** a definition of the binary file format on disk.
+
+This RFC is mainly addressing the logical order of protobuf messages within a
+decoded binary stream (where decoded e.g. refers to the decompressed message
+stream for *pbstream* files).
+
+## Motivation
+
+[motivation]: #motivation
+
+At the moment, the file format for serializing the state of Cartographer is
+defined purely within the serialization & deserialization routines of
+Cartographer. Since we serialize all information into a binary stream of
+`protobuf` messages (*pbstream*) without a format-version information, extending
+or changing the format inevitably leads to a breaking change for Cartographer
+users, as there is no way of falling back to load previously serialized states.
+
+This RFC is going to address this issue by introducing a `SerializationHeader`
+message and an interface for parsing serialized states, independent of the
+serialization format.
+
+## Approach
+
+[approach]: #approach
+
+For being able to identify the underlying serialization format, we introduce a
+`SerializationHeader` message.
+
+```
+message SerializationHeader {
+  uint32 format_version = 1;
+}
+```
+
+We further change the current `SerializedData` protobuf message, which is meant
+to be able to contain `oneof` any of the message types required for serializing
+the mapping state:
+
+```
+message SerializedData {
+  oneof data {
+    PoseGraph pose_graph = 1;
+    AllTrajectoryBuilderOptions all_trajectory_builder_options = 2;
+    Submap submap = 3;
+    TrajectoryData trajectory_data = 4;
+    Node node = 5;
+    ImuData imu_data = 6;
+    OdometryData odometry_data =7;
+    FixedFramePoseData fixed_frame_pose_data = 8;
+    LandmarkData landmark_data = 9;
+  }
+}
+```
+
+~~For backwards compatibility, we will rename the original `SerializedData`
+definition to `LegacySerializedData`.~~ (see discussion)
+
+By keeping all the payload related data wrapped in a single message type, we can
```

Comment:
+1. I think riegeli did not exist when we added the pbstream format. Would be great to remove the pbstream code and use this library instead. This would also allow us to change to faster compression.